lang/funcs/collection.go

```diff
@@ -106,6 +107,145 @@ var LengthFunc = function.New(&function.Spec{
+// and returns the first one that isn't empty.
+var CoalesceListFunc = function.New(&function.Spec{
+	Params: []function.Parameter{},
+	VarParam: &function.Parameter{
+		Name:             "vals",
+		Type:             cty.DynamicPseudoType,
+		AllowUnknown:     true,
+		AllowDynamicType: true,
+		AllowNull:        true,
+	},
+	Type: func(args []cty.Value) (ret cty.Type, err error) {
+		if len(args) == 0 {
+			return cty.NilType, fmt.Errorf("at least one argument is required")
+		}
+
+		argTypes := make([]cty.Type, len(args))
+
+		for i, arg := range args {
+			argTypes[i] = arg.Type()
+		}
+
+		retType, _ := convert.UnifyUnsafe(argTypes)
+		if retType == cty.NilType {
+			return cty.NilType, fmt.Errorf("all arguments must have the same type")
+		}
+
+		return retType, nil
+	},
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		vals := make([]cty.Value, 0, len(args))
+		for _, arg := range args {
+
+			// We already know this will succeed because of the checks in our Type func above
+			arg, _ = convert.Convert(arg, retType)
+
+			it := arg.ElementIterator()
+			for it.Next() {
+				_, v := it.Element()
+				vals = append(vals, v)
+			}
+
+			if len(vals) > 0 {
+				return cty.ListVal(vals), nil
+			}
+		}
+
+		return cty.NilVal, fmt.Errorf("no non-null arguments")
+	},
+})
+
+// CompactFunc contructs a function that takes a list of strings and returns a new list
+// with any empty string elements removed.
+var CompactFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.String),
+		},
+	},
+	Type: function.StaticReturnType(cty.List(cty.String)),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		var outputList []cty.Value
+
+		for it := args[0].ElementIterator(); it.Next(); {
+			_, v := it.Element()
+			if v.AsString() == "" {
+				continue
+			}
+			outputList = append(outputList, v)
+		}
+		return cty.ListVal(outputList), nil
+	},
+})
+
+// ContainsFunc contructs a function that determines whether a given list contains
+// a given single value as one of its elements.
+var ContainsFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.DynamicPseudoType),
+		},
+		{
+			Name: "value",
+			Type: cty.DynamicPseudoType,
+		},
+	},
+	Type: function.StaticReturnType(cty.Bool),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		_, err = Index(args[0], args[1])
+		if err != nil {
+			fmt.Println(err)
+			return cty.False, nil
+		}
+
+		return cty.True, nil
+	},
+})
+
+// IndexFunc contructs a function that finds the element index for a given value in a list.
+var IndexFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.DynamicPseudoType,
+		},
+		{
+			Name: "value",
+			Type: cty.DynamicPseudoType,
+		},
+	},
+	Type: function.StaticReturnType(cty.Number),
```

Comment:
Unfortunately in this case we'll need a hand-written type check function to ensure that `args[0]` is a list or tuple. If not, then passing in other types will have weird effects:

* Passing in a primitive type will panic, because `LengthInt()` doesn't apply to primitive types.
* Passing in a map will cause a strange error (possibly also a panic?) because `LengthInt` and `ElementIterator` both work for maps but the "index" returned will be a string key, and thus it will not agree with what was returned by this type function.

Looking at the implementation again, I think it should work with a simple test like `args[0].Type().IsListType() || args[0].Type().IsTupleType()`, returning an error if not, because the implementation function is already compatible with both types, and the return type is still "static".